File: python/leetcode/BinaryTreeDuplicate.py
'''
Leetcode problem No 652 Find Duplicate Subtrees
Solution written by Xuqiang Fang on 13 July, 2018
'''

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        counts = {}
        ans = []
        self.serialize(root, counts, ans)
        return ans

    def serialize(self, root, counts, ans):
        if not root:
            return '#'
        key = str(root.val) + ',' + self.serialize(root.left, counts, ans) \
            + ',' + self.serialize(root.right, counts, ans)
        if key not in counts:
            counts[key] = 0
        counts[key] += 1
        if counts[key] == 2:
            print('here')
            ans.append(root)
        return key

    def findDuplicateSubtrees_(self, root):
        ##second method
        ids = {}; counts = {}
        ans = []
        self.get(root, ids, counts, ans)
        return ans

    def get(self, root, ids, counts, ans):
        if not root:
            return 0
        key = (root.val << 32) | (self.get(root.left, ids, counts, ans) << 16) \
            | (self.get(root.right, ids, counts, ans))
        if key not in ids:
            ids[key] = len(ids) + 1
        k = ids[key]
        if k not in counts:
            counts[k] = 0
        counts[k] += 1
        if counts[k] == 2:
            ans.append(root)
        return k

def print_tree(root):
    if root:
        print_tree(root.left)
        print_tree(root.right)
        print(root.val)

def main():
    s = Solution()
    r = TreeNode(1)
    a = TreeNode(2)
    b = TreeNode(3)
    c = TreeNode(4)
    d = TreeNode(2)
    e = TreeNode(4)
    f = TreeNode(4)
    r.left = a
    r.right = b
    a.left = c
    b.left = d
    b.right = e
    d.left = f
    print_tree(a)
    ans = s.findDuplicateSubtrees_(r)
    print('length is %d'%len(ans))
    for i in ans:
        print_tree(i)

if __name__ == '__main__':
    main()

File: java/leetcode/BinaryTreeDuplicate.java
//Leetcode problem 652 Find Duplicate Subtrees
//Solution written by Xuqiang Fang on 13 July, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class TreeNode{
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){val = x;}
}
class Solution{
    public List<TreeNode> findDuplicateSubtrees_(TreeNode root){
        Map<String, Integer> map = new HashMap<>();
        List<TreeNode> ans = new ArrayList<>();
        dfs(root, map, ans);
        return ans;
    }
    private String dfs(TreeNode root, Map<String, Integer>map, List<TreeNode> ans){
        if(root == null){
            return "#";
        }
        String key = root.val + "," + dfs(root.left, map, ans) + "," + dfs(root.right, map, ans);
        map.put(key, map.getOrDefault(key, 0)+1);
        if(map.get(key) == 2){
            ans.add(root);
        }
        return key;
    }

    public List<TreeNode> findDuplicateSubtrees(TreeNode root){
        //second method
        Map<Long, Integer> ids = new HashMap<>();
        Map<Long, Integer> counts = new HashMap<>();
        List<TreeNode> ans = new ArrayList<>();
        getId(root, ids, counts, ans);
        return ans;
    }
    private long getId(TreeNode root, Map<Long, Integer> ids, Map<Long, Integer> counts, List<TreeNode> ans){
        if (root == null){
            return 0;
        }
        //the basic idea is the same, give each subtree an id and if two subtrees are the same
        //they should have the same id, so the id has to do with node values
        long key = ((long)root.val << 31)|(getId(root.left,ids,counts,ans)<<15)|(getId(root.right,ids,counts, ans));
        int id;
        if (!ids.containsKey(key)){
            int size = ids.size();
            ids.put(key, size+1); 
            id = size + 1;
        }
        else{
            id = ids.get(key);
        }
        counts.put(key, counts.getOrDefault(key, 0) + 1);
        if(counts.get(key) == 2){
            ans.add(root);
        }
        return id;
    }
}

public class BinaryTreeDuplicate{
	public static void main(String[] args){
		Solution s = new Solution();
        TreeNode root = new TreeNode(1);
        TreeNode a = new TreeNode(2);
        TreeNode b = new TreeNode(3);
        TreeNode c = new TreeNode(4);
        TreeNode d = new TreeNode(2);
        TreeNode e = new TreeNode(4);
        TreeNode f = new TreeNode(4);
        root.left = a;
        root.right = b;
        a.left = c;
        b.left = d;
        b.right = e;
        d.left = f;
        List<TreeNode> ans = s.findDuplicateSubtrees(root);
        for(TreeNode t : ans){
            print(t);
        }
	}
    public static void print(TreeNode r){
        if(r != null){
            print(r.left);
            print(r.right);
            System.out.println(r.val);
        }
    }
}
